File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private static final int[] dx = {-1, 0, 1, 0, -1, 1, 1, -1, -2, 0, 2, 0};
    private static final int[] dy = {0, 1, 0, -1, 1, 1, -1, -1, 0, 2, 0, -2};
    
    private class Point {
        private final int x;
        private final int y;
        
        private Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
        
        public int getX() {
            return x;
        }
        
        public int getY() {
            return y;
        }
    }
    
    private boolean isDistanced(char[][] place, Point p) {
        int x = p.getX();
        int y = p.getY();

        for (int i = 0; i < 12; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx < 0 || nx >= 5 || ny < 0 || ny >= 5 || place[nx][ny] != 'P') {
                continue;
            }

            int distX = Math.abs(nx - x);
            int distY = Math.abs(ny - y);

            int mx = (nx < x) ? x - 1 : x + 1;
            int my = (ny < y) ? y - 1 : y + 1;

            if (distX == 2) {
                if (place[mx][ny] == 'O') {
                    return false;
                }
            } else if (distY == 2) {
                if (place[nx][my] == 'O') {
                    return false;
                }
            } else if (distX == 1 && distY == 1) {
                if (!(place[nx][y] == 'X' && place[x][ny] == 'X')) {
                    return false;
                }
            } else if (place[nx][ny] == 'P') {
                return false;
            }
        }
        
        return true;
    }
    
    private boolean isPossiblePlace(char[][] place) {
        List<Point> persons = new ArrayList<>();
        
        for (int x = 0; x < 5; x++) {
            for (int y = 0; y < 5; y++) {
                if (place[x][y] == 'P') {
                    persons.add(new Point(x, y));
                }
            }
        }
        
        for (Point person : persons) {
            if (!isDistanced(place, person)) {
                return false;
            }
        }
        
        return true;
    }
    
    public int[] solution(String[][] places) {
        int[] answer = new int[places.length];
        
        for (int i = 0; i < places.length; i++) {
            char[][] place = new char[5][];
            for (int j = 0; j < 5; j++) {
                place[j] = places[i][j].toCharArray();
            }
            
            if (isPossiblePlace(place)) {
                answer[i] = 1;
            } else {
                answer[i] = 0;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
def solution(places):
    dx = (-1, 0, 1, 0, -1, 1, 1, -1, -2, 0, 2, 0)  
    dy = (0, 1, 0, -1, 1, 1, -1, -1, 0, 2, 0, -2)
    
    return [check_place(p, dx, dy) for p in places]


def check_place(place, dx, dy):
    p_points = [(x, y) for x in range(5) for y in range(5) if place[x][y] == 'P']
    
    for x, y in p_points:
        for i in range(12):
            nx, ny = x + dx[i], y + dy[i]
            if nx < 0 or nx >= 5 or ny < 0 or ny >= 5 or place[nx][ny] != 'P':
                continue

            dist_x, dist_y = abs(nx - x), abs(ny - y)

            # x축으로 두 칸 이동
            if dist_x == 2:
                if place[x - 1 if nx < x else x + 1][ny] == 'O':
                    return 0
            # y축으로 두 칸 이동
            elif dist_y == 2:
                if place[nx][y - 1 if ny < y else y + 1] == 'O':
                    return 0
            # 대각선으로 한 칸 이동
            elif dist_x == dist_y == 1:
                if not place[nx][y] == place[x][ny] == 'X':
                    return 0
            # 한 칸 이동 시
            elif place[nx][ny] == 'P':
                return 0
            
    return 1
